### PR TITLE
fix: Remove DSCI requirement for service reconciliation

### PR DESCRIPTION
**Pull request by ugiordan**
## Description

This PR addresses [RHOAIENG-18035](https://issues.redhat.com/browse/RHOAIENG-18035) by removing the hard requirement for DSCInitialization (DSCI) to exist when reconciling services.

### Problem
The reconciler was originally written for components and assumed DSCI must always be present. This caused services (like Auth, Gateway, Monitoring) to fail reconciliation and remain stuck in a "NotReady" state when DSCI didn't exist yet, only recovering after backoff retry.

### Solution
**Core Changes:**
- **Reconciler**: Modified to distinguish between "DSCI not found" (acceptable) vs actual errors (fail). Services now reconcile successfully with `rr.DSCI = nil` when DSCI doesn't exist.
- **Hash function**: Added nil check to prevent panic when hashing reconciliation requests without DSCI.
- **ApplicationNamespace helper**: Implemented two-tier fallback - uses cached DSCI if available, otherwise attempts on-demand fetch.

**Service-Specific Updates:**
- **Auth service**: Updated `bindRole` function to use `ApplicationNamespace()` helper instead of direct DSCI access.
- **Monitoring service**: 
  - Updated template data to use `ApplicationNamespace()` helper
  - Added nil check in `cleanupPrometheusRules` to skip cleanup when DSCI unavailable
- **Templates**: Updated Auth template to handle nil DSCI gracefully.

**JIRA:** https://issues.redhat.com/browse/RHOAIENG-18035

## How Has This Been Tested?

- ✅ `make build` - Compilation successful
- ✅ `make lint` - No linting errors in modified files
- ✅ `make test` - Existing test suite passes

All modified files pass linting, and existing tests continue to work, verifying backward compatibility.

## Screenshot or short clip

N/A - Internal reconciliation logic change with no UI impact.

## Merge criteria

- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [x] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification

This PR modifies internal reconciliation logic to handle missing DSCI gracefully. The changes are backward compatible, and all existing E2E tests with DSCI present will continue to pass unchanged. The new behavior (handling missing DSCI) is covered by unit tests and doesn't introduce new features requiring E2E coverage. Services function identically when DSCI exists; they just handle the edge case of missing DSCI gracefully instead of failing.

## Summary by CodeRabbit

* **Refactor**
  * Controllers and runtime flows now resolve namespaces at runtime from cluster; reconciliation runs actions sequentially, records provisioning status, and stops on first failure.

* **Templates**
  * Rendered templates and manifests use a unified AppNamespace value for consistent resource scoping.

* **Documentation**
  * Monitoring CRD clarified: collectorReplicas default is 1 on single-node and 2 on multi-node clusters.

* **Tests**
  * Consolidated fake-client test helpers; tests now source namespace data externally rather than embedding it.

### CODE REVIEW

File: pkg/controller/reconciler/reconciler.go [hunk 1]
```diff
@@ -291,11 +291,20 @@ func (r *Reconciler) apply(ctx context.Context, res common.PlatformObject) error
...
-	case dscilErr != nil:
+	case dscilErr != nil && !k8serr.IsNotFound(dscilErr):
+		// Failed to fetch DSCI due to an actual error (e.g., API server issues, permissions).
+		// This is different from DSCI simply not existing yet, which is handled below.
 		provisionErr = fmt.Errorf("failed to get DSCInitialization: %w", dscilErr)
 	default:
+		// DSCI either exists or doesn't exist (NotFound). Both cases are acceptable.
+		// Services can check rr.DSCI != nil before accessing DSCI-specific fields.
+		if dscilErr == nil {
+			rr.DSCI = dsci.DeepCopy()
```
Comment (lburgazzoli) [inline pkg/controller/reconciler/reconciler.go]: I'd remove rr.DSCI entirely

Comment (ugiordan) [reply pkg/controller/reconciler/reconciler.go]: We should no longer fetch the `DSCI,` right? In that case, it should be removed from the `ReconciliationRequest`, and the reconcilers updated as needed.

File: internal/controller/services/auth/auth_controller_actions.go [hunk 2]
```diff
@@ -55,6 +55,12 @@ func initialize(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
 }
 
 func bindRole(ctx context.Context, rr *odhtypes.ReconciliationRequest, groups []string, roleBindingName string, roleName string) error {
+	namespace, err := actions.ApplicationNamespace(ctx, rr)
```
Comment (davidebianchi) [inline internal/controller/services/auth/auth_controller_actions.go]: In cases where services uses applicationNamespace, what happens if namespace changes in DSCI? Shouldn't services which needs application namespace watch DSCI?

Comment (ugiordan) [reply internal/controller/services/auth/auth_controller_actions.go]: Yes, they have to watch for the DSCI changes. I'll update that once I understand https://github.com/opendatahub-io/opendatahub-operator/pull/2750#discussion_r2455727547.

Comment (lburgazzoli) [reply internal/controller/services/auth/auth_controller_actions.go]: btw `applicationNamespace` never changes, it is essentially read-only and defined at DSCI creation time

File: internal/controller/services/monitoring/monitoring_controller_support.go [hunk 3]
```diff
@@ -223,12 +224,17 @@ func getTemplateData(ctx context.Context, rr *odhtypes.ReconciliationRequest) (m
 		return nil, errors.New("instance is not of type services.Monitoring")
 	}
 
+	appNamespace, err := actions.ApplicationNamespace(ctx, rr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get applications namespace: %w", err)
+	}
```
Comment (zdtsw) [inline internal/controller/services/monitoring/monitoring_controller_support.go:229]: just `return nil, err` ?
if error has been formated within `ApplicationNamespace()`

File: internal/controller/services/auth/auth_controller_actions.go [hunk 4]
```diff
@@ -55,6 +55,12 @@ func initialize(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
...
+	if err != nil {
+		logf.Log.Error(err, "error getting application namespace")
+		return err
+	}
```
Comment (zdtsw) [inline internal/controller/services/auth/auth_controller_actions.go:61]: similar here?

File: pkg/controller/actions/actions_support.go [hunk 5]
```diff
@@ -2,13 +2,27 @@ package actions
...
+	if rr.DSCI != nil {
+		return rr.DSCI.Spec.ApplicationsNamespace, nil
+	}
+
+	// DSCI was not available during reconciliation setup (e.g., service started before
+	// platform initialization). Attempt an on-demand fetch to support actions that
+	// require ApplicationsNamespace. If DSCI truly doesn't exist, the error will
+	// propagate and the caller must handle appropriately.
+	dsci, err := cluster.GetDSCI(ctx, rr.Client)
+	if err != nil {
```
Comment (zdtsw) [inline pkg/controller/actions/actions_support.go]: this should be if `k8serr.IsNotFound(err) { `, right?

File: internal/controller/components/dashboard/dashboard_controller_actions.go [hunk 6]
```diff
@@ -71,19 +72,25 @@ func setKustomizedParams(ctx context.Context, rr *odhtypes.ReconciliationRequest
...
 
-func configureDependencies(_ context.Context, rr *odhtypes.ReconciliationRequest) error {
+func configureDependencies(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
 	if rr.Release.Name == cluster.OpenDataHub {
 		return nil
 	}
 
-	err := rr.AddResources(&corev1.Secret{
+	// Fetch application namespace from DSCI.
+	appNamespace, err := deployments.ApplicationNamespace(ctx, rr)
```
Comment (lburgazzoli) [inline internal/controller/components/dashboard/dashboard_controller_actions.go]: maybe move the `ApplicationNamespace` function to another package, like `cluster.ApplicationNamespace(ctx, cli)` ?

File: internal/controller/services/monitoring/monitoring_controller_support.go (hunk 3 above)
Comment (ugiordan) [reply internal/controller/services/monitoring/monitoring_controller_support.go:229]: Good catch. Thx.

File: internal/controller/services/auth/auth_controller_actions.go (hunk 4 above)
Comment (ugiordan) [reply internal/controller/services/auth/auth_controller_actions.go:61]: Done.

File: internal/controller/components/dashboard/dashboard_controller_actions.go (hunk 6 above)
Comment (ugiordan) [reply internal/controller/components/dashboard/dashboard_controller_actions.go]: Yep.

**Comment by ugiordan** [comment]
@grdryn, should we remove this line? https://github.com/opendatahub-io/opendatahub-operator/blob/f97ac9b1760693944540948bc403a9b1de575334/internal/controller/components/kserve/kserve_controller.go#L114-L115?

**Comment by ugiordan** [comment]
I want to get https://github.com/opendatahub-io/opendatahub-operator/pull/2692 merged first before cherry-picking.
